---
title: 'Deconvolution with scaden: SRP058181'
author: 
- name: "Regina H. Reynolds"
  affiliation: UCL
output: 
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}

library(biomaRt)
library(DescTools)
library(DT)
library(DESeq2)
library(factoextra)
library(ggsci)
library(ggpubr)
library(readxl)
library(tidyverse)
library(stringr)

# Themes
theme_rhr <-  theme_bw(base_family = "Helvetica") + 
  theme(panel.grid.major.x = element_blank(),
        legend.position = "top",
        strip.text = element_text(size = 8),
        axis.text.x = element_text(size = 8, angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(size = 8),
        axis.title.y = element_text(vjust = 0.6),
        axis.title = element_text(size = 10),
        panel.spacing = unit(0.1, "lines"))

theme_rhr_coord_flip <-  theme_bw(base_family = "Helvetica") + 
  theme(panel.grid.major.y = element_blank(),
        legend.position = "top",
        strip.text = element_text(size = 8),
        axis.text.x = element_text(size = 8),
        axis.text.y = element_text(size = 8),
        axis.title.y = element_text(vjust = 0.6),
        axis.title = element_text(size = 10),
        strip.text.y = element_text(angle = 90),
        panel.spacing = unit(0.1, "lines"))

knitr::opts_chunk$set(echo = F, warning = F, message= F)

```

> Aim: deconvolution of SRP058181 dataset using Scaden and our own single-nucleus RNA-sequencing data.

# File paths/files for workflow

```{r file-paths, echo = T}

source(here::here("R", "file_paths.R"))
source(here::here("R", "biomart_df.R"))
source(here::here("R", "scaden_rankings_and_calculate_ratios.R"))


sample_info <- 
  readRDS(
    file = 
      file.path(
        path_to_results,
        "SRP058181/gene_level_quant/SRP058181_DESeqDataSet.Rds"
      )
    ) %>% 
  colData() %>% 
  as_tibble() %>% 
  # Remove SRR2015746 (C0061), due to uncertainty re. sex
  dplyr::filter(recount_id != "SRR2015746")

```

# Implementing Scaden
- Will need to conduct all three steps below:
    1. Training data generation. Does not need to be re-run.
    2. Pre-processing. Despite having run this previously with our own data, this will need to be re-run, as there may be some genes present in SRP058181 that were not present within our own data. And to run, scaden, must have the same genes present in training data and bulk data.
    2. Training of 3 deep neural network models.
    3. Prediction of cell-type proportions in bulk tissue.

## Pre-processing bulk data
```{r, echo = T,  eval = F}
dds <- 
  readRDS(
    file = 
      file.path(
        path_to_results,
        "SRP058181/gene_level_quant/SRP058181_DESeqDataSet.Rds"
      )
    )

# Remove SRR2015746 (C0061), due to uncertainty re. sex
dds <- dds[, dds$recount_id != "SRR2015746"]

# Library size matrix
lib_matrix <- colSums(counts(dds)) %>% 
    as.matrix() %>% 
    t()

lib_matrix <- lib_matrix[rep(1, times = nrow(counts(dds))),]

# Divide raw counts by library size and scale with 1E6 to get counts per million
norm_counts <- (counts(dds)/lib_matrix) * 1000000

# Processing for final scaden format
# Remove ".*" following each ensembl ID
# Convert ensembl IDs to gene symbols
norm_counts <- norm_counts %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "gene") %>% 
  dplyr::mutate(gene = str_replace(gene, "\\..*", "")) %>% 
  # According to recount2, study aligned to hg38
  biomart_df(
    columnToFilter = "gene", 
    mart = 38, 
    attributes = c("ensembl_gene_id", "hgnc_symbol"), 
    filter = "ensembl_gene_id"
    ) %>% 
  dplyr::select(gene, hgnc_symbol, everything())

# Filter out rows where ensembl ID has no corresponding gene symbol
# Remove duplicate hgnc_symbols (i.e. multiple ensembl ids mapping to same hgnc_symbol)
norm_counts <- norm_counts %>% 
  dplyr::filter(hgnc_symbol != "") %>%
  dplyr::filter(!(duplicated(hgnc_symbol) | duplicated(hgnc_symbol, fromLast = TRUE))) %>% 
  dplyr::select(-gene)

# Save for scaden
write_delim(
  format(norm_counts, scientific = FALSE),
  path = 
    file.path(
      path_to_raw_data, 
      "deconvolution/scaden/SRP058181_norm_counts.txt"
    ), 
  delim = "\t"
)
```

## Bulk simulation, creation of h5ad file, training and prediction
- Ran with:
    - Cells: 1000
    - Samples: 1000
    - Training steps: 5000
- Choice of cells and samples based on optimisation performed in [deconvolution_scaden.html](../deconvolution/deconvolution_scaden.html).
```{bash echo = T, eval = F}

# copy files into container from server
docker cp /home/rreynolds/projects/Aim2_PDsequencing_wd/raw_data/deconvolution/scaden/SRP058181_norm_counts.txt scaden_rreynolds:/home/data/SRP058181/

# Re-initiate docker
docker exec -it scaden_rreynolds bash

# Make directory
mkdir /home/data/SRP058181/cells.1000_samples.1000/

# Run scaden 
scaden process /home/data/PD_seq/scaden_2020Feb/cells.1000_samples.1000/pdseq_bulksim.h5ad /home/data/SRP058181/SRP058181_norm_counts.txt --processed_path /home/data/SRP058181/cells.1000_samples.1000/processed_pdseq_bulksim.h5ad

scaden train /home/data/SRP058181/cells.1000_samples.1000/processed_pdseq_bulksim.h5ad --model_dir /home/data/SRP058181/cells.1000_samples.1000/scaden_model_5000_steps --steps 5000

scaden predict /home/data/SRP058181/SRP058181_norm_counts.txt --model_dir /home/data/SRP058181/cells.1000_samples.1000/scaden_model_5000_steps --outname /home/data/SRP058181/cells.1000_samples.1000/scaden_predictions_cells.1000:samples.1000.txt

# Copy docker output to server
docker cp scaden_rreynolds:/home/data/SRP058181/cells.1000_samples.1000/scaden_predictions_cells.1000:samples.1000.txt /home/rreynolds/projects/Aim2_PDsequencing_wd/results/deconvolution/scaden/SRP058181/scaden_predictions_cells.1000:samples.1000.txt 

```

```{r, eval = F, echo = T}
## Check number of genes in common between training data and bulk data (see dataset.dims in .h5ad object X)

library(hdf5r)

# docker cp scaden_rreynolds:/home/data/SRP058181/cells.1000_samples.1000/processed_pdseq_bulksim.h5ad /home/rreynolds/downloads/SRP058181_processed_pdseq_bulksim.h5ad

file.h5 <- H5File$new("/home/rreynolds/downloads/SRP058181_processed_pdseq_bulksim.h5ad", mode="r+")
file.h5$ls(recursive=TRUE) # 14094 x 24000
file.h5$close_all()

```

# Results
```{r}

# Read in new cell type labels
ct <- 
  read_delim(file = 
               file.path(
                 path_to_raw_data,
                 "Cell.Type.Annotations.For.All.Samples.csv"
               ), 
             delim = " ", 
             col_names = c("ID", "Celltype"), 
             skip = 1)

# Calculate cell type proportions from cell type assignments
sn_proportions <- 
  ct %>%
  dplyr::mutate(Celltype = case_when(Celltype %in% c("Endo", "Per") ~ "Vascular", 
                                     TRUE ~ Celltype)) %>% 
  tidyr::separate(ID, into = c("sample_id", "disease_group", "ID"), sep = "_") %>% 
  dplyr::group_by(Celltype, sample_id) %>% 
  dplyr::summarise(n = n()) %>% 
  dplyr::group_by(sample_id) %>% 
  dplyr::mutate(snRNA_proportions = n/sum(n)) %>% 
  dplyr::ungroup() %>% 
  dplyr::select(-n) %>% 
  tidyr::spread(key = Celltype, value = snRNA_proportions)

# Create dataframe with cell type classes
cell_type_class <- 
  data.frame(
    Celltype = sn_proportions %>% 
      dplyr::select(-sample_id) %>% 
      colnames() %>% 
      sort(),
    Celltype_class = factor(c("Astrocyte", "Neuron", "Neuron", 
                              "Immune", "Oligo", "Oligo", "Vascular"),
                            levels=c("Astrocyte", "Immune", "Oligo", "Vascular", 
                                     "Neuron"))
  )

# Read in scaden predictions, determine within-sample rankings of cell-types by proportions, and calculate glia-to-neuron and non-neuron-to-neuron ratios
results <- 
  scaden_rankings_and_calculate_ratios(
    prediction_dir = 
      file.path(
        path_to_results,
        "deconvolution/scaden/SRP058181/"
      ),
    unclassified_cell_type = "Unclassified", 
    sample_info = sample_info %>% 
      dplyr::select(-sample_id) %>% 
      dplyr::rename(sample_id = recount_id), 
    snRNAseq_proportions = NULL, 
    cell_type_class = cell_type_class,
    glia = vars(Astro, Microglia, Oligo, OPC), 
    nonneuron = vars(Astro, Vascular, Microglia, Oligo, OPC), 
    neuron = vars(Excitatory, Inhibitory),
    astrocyte = vars(Astro),
    oligos = vars(Oligo, OPC),
    microglia = vars(Microglia)
  )

# saveRDS(results, file.path(path_to_results, "deconvolution/scaden/SRP058181/scaden_summary_results.Rds"))

```

## Cell-type proportions across disease groups
```{r echo = F, eval = T, fig.height = 8, fig.cap="Cell-type proportions of SRP058181 derived from scaden deconvolution predictions grouped by disease status. Scaden neural networks were trained using simulated bulk RNA-sequencing profiles generated from each individual's single-nuclear RNA-sequencing data; 1000 simulated bulk samples were prepared per subject, using 1000 nuclei per sample. Once trained, Scaden was used to predict cell-type proportions for bulk RNA-sequencing data derived from the same individuals. Single-nuclear cell-type proportions were calculated per individual, by dividing the number of cells assigned to a cell type by the total number of cells. Cell-type proportions are summarised by cell type in each disease group. Cell types are coloured by their overarching cell-type class."}

results$ranks %>% 
  ggplot(
    aes(
      x = fct_relevel(Celltype,
                      results$ranks %>% 
                        dplyr::filter(
                          source == "Scaden deconvolution" & cells == 100 & samples == 1000
                        ) %>%  
                        arrange(Celltype_class, Celltype) %>% 
                        .[["Celltype"]] %>% 
                        unique() %>% 
                        as.character()), 
      y = cell_type_proportion, 
      fill = Celltype_class
    )
  ) + 
  geom_boxplot() +
  facet_grid(~ Disease_Group) +
  ylim(0,100) +
  scale_fill_manual(values = pal_npg("nrc", alpha = 1)(10)[c(4,2,1,6,3,10)]) +
  labs(x = "", y = "Cell type proportion", fill = "Cell-type class") + 
  theme_rhr 

```
- Looking at control population, excitatory neurons come out the top population, followed by oligodendrocytes. 
- Within glial populations, as expected, oligodendrocytes rank first, followed by astrocytes and finally microglia.
- Notably, oligodendrocyte proportions vary a lot over each of the disease conditions (much more so than what was observed in our own data, although that may be due to the small sample sizes per disease group), with some samples almost entirely made up of oligodendrocytes.
- Worth noting that spread of data does not appear to be determined by absolute nuclei numbers from snRNA-seq data i.e. nuclei assigned to excitatory > nuclei assigned to microglia > nuclei assigned to vascular, yet spread is greatest for excitatory and vascular as compared to microglia. 

## Glia-to-neuron ratios
```{r glia-neuron-ratio, echo = F, fig.cap="Ratio of glia-to-neurons and non-neurons-to-neurons across disease groups, as calculated using cell-type proportions derived from scaden deconvolution predictions or single-nuclear cell-type assignments. Scaden neural networks were trained using simulated bulk RNA-sequencing profiles generated from each individual's single-nuclear RNA-sequencing data; 1000 simulated bulk samples were prepared per subject, using 100 nuclei per sample."}

results$ratios %>% 
  ggplot(aes(x = Disease_Group, y = value)) + 
  geom_boxplot(aes(fill = ratio)) + 
  geom_hline(yintercept = 1.5, linetype = 2, colour = "black") + 
  facet_wrap(~ source) +
  labs(x = "", y = "Value of ratio", fill = "Type of ratio") + 
  coord_cartesian(ylim = c(0,50)) +
  theme_rhr

```

```{r}
# Median values
results$ratios %>%
  dplyr::group_by(source, Disease_Group, ratio) %>%
  dplyr::summarise(median = median(value))
```
- Median glia:neuron ratio for controls calculated using scaden predictions is 1.18, which is very close to the expected ~ 1.5. 

## Cell-type proportions relative to control
```{r echo = F, eval = T, fig.height = 5, fig.cap="Cell-type proportions derived from scaden deconvolution. Scaden neural networks were trained using simulated bulk RNA-sequencing profiles generated from each individual's single-nuclear RNA-sequencing data; 1000 simulated bulk samples were prepared per subject, using 1000 nuclei per sample. Once trained, Scaden was used to predict cell-type proportions for bulk RNA-sequencing data derived from the same individuals. Cell-type proportions are grouped by cell type and disease status and displayed relative to the median of controls (within a cell type)."}

results$ranks %>%
  dplyr::select(sample_id, Celltype, cell_type_proportion, Celltype_class, Disease_Group) %>% 
  tidyr::spread(key = Disease_Group, value = cell_type_proportion) %>% 
  dplyr::group_by(Celltype) %>%
  dplyr::arrange(Celltype) %>% 
  dplyr::mutate(median_control = median(Control, na.rm = T)) %>%
  dplyr::ungroup() %>% 
  # Calculate fold change from reference, controls, by dividing each column by median_control
  dplyr::mutate_at(.vars = vars("Control", "PD", "PDD"), .funs = function(x){x/.[["median_control"]]}) %>%
  dplyr::select(-median_control) %>% 
  tidyr::gather(
    key = "Disease_Group", 
    value = "cell_type_proportion_relative_to_control", 
    -sample_id, -Celltype, -Celltype_class
    ) %>% 
  dplyr::filter(!is.na(cell_type_proportion_relative_to_control)) %>% 
  # dplyr::filter(Celltype == "Excitatory") %>% 
  ggplot(aes(x = Celltype, 
             y = cell_type_proportion_relative_to_control, 
             fill = fct_relevel(Disease_Group,
                                c("Control", "PD", "PDD")))) + 
  geom_boxplot() +
  scale_fill_manual(values = c("#868686", "#c4ebe3", "#73cab9", "#19967d")) +
  labs(x = "Cell type", y = "Cell type proportion relative to controls", fill = "Disease group") + 
  theme_rhr

```

### Kruskal test
```{r}
results$ranks %>%
  dplyr::filter(Celltype!= "Unclassified") %>% 
  dplyr::group_by(Celltype) %>% 
  dplyr::do(kruskal.test(x= .$cell_type_proportion, g = .$Disease_Group) %>% 
              broom::tidy()) %>%
  dplyr::arrange(p.value) %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')
```


### Wilcoxon rank-sum test results
- Use Wilcoxon rank-sum test (non-parametric test), due to non-normal distribution. Also tends to have more power when data contains extreme outliers. 
    - FDR correction per cell type i.e. correcting for number of comparisons between disease groups (6). Significance is FDR < 0.05, and nominal significance is FDR < 0.1.    
    - Astrocytes have significantly higher proportions in PD compared to controls. 
    - Excitatory neurons have a nominally significant lower proportion in PD and PDD compared to controls (PD, FDR = 0.056; PDD, FDR = 0.060). Inhibitory neurons found to have significantly lower proportions in PD and PDD compared to controls. *In our own data, no significant differences were found between disease groups.*
    - Microglia have significantly higher proportions in PD compared to controls. Also, microglia have a nominally significant higher proportion in PDD compared to controls (FDR = 0.096). *In our own data, microglia had nominally significantly higher proportion across all disease states compared to control.*
    - OPCs have nominally significant higher proportions in PDD compared to controls and PD (control, FDR = 0.11; PD, FDR = 0.11). *In our own data, OPCs had significantly higher proportions in DLB compared to control, PD and PDD. Also, nominally significant higher proportions of OPCs in PDD compared to control (p = 0.052).*
    - Vascular cells have nominally signifcant higher proportions in PD compared to control (FDR = 0.054). *In our own data, vascular cells had significantly higher proportions in DLB compared to control, PD and PDD. Also, nominally significant higher proportions in PD and PDD compared to control.*
```{r echo = T}
results$ranks %>%
  dplyr::filter(Celltype!= "Unclassified") %>% 
  dplyr::group_by(Celltype) %>% 
  dplyr::do(pairwise.wilcox.test(x= .$cell_type_proportion, g = .$Disease_Group, p.adjust.method = "none", paired = FALSE) %>% 
              broom::tidy()) %>%
  dplyr::mutate(fdr = p.adjust(p.value, method = "BH")) %>%
  dplyr::arrange(fdr) %>% 
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```

# Comparison to deconvolution of own data

## Qualitative comparison
```{r, fig.height = 6, fig.cap="Cell-type proportions derived from own dataset or SRP058181. For both datasets, cell-type proportions are grouped by cell type and coloured by disease status."}

results$ranks %>%
  dplyr::select(sample_id, Celltype, cell_type_proportion, Celltype_class, Disease_Group) %>% 
  dplyr::mutate(dataset = "SRP058181") %>% 
  dplyr::bind_rows(
    read_delim(
      file = 
        file.path(
          path_to_results,
          "deconvolution/scaden/2020Feb/scaden_summary_results.txt"),
      delim = "\t"
    ) %>% 
      dplyr::filter(
        source == "Scaden deconvolution", cells == 1000, samples == 1000, Disease_Group != "DLB"
      ) %>% 
      dplyr::select(
        sample_id, Celltype, cell_type_proportion, 
        Celltype_class, Disease_Group
      ) %>% 
      dplyr::mutate(dataset = "Own")
  ) %>% 
  ggplot(
    aes(
      x = Celltype, 
      y = cell_type_proportion, 
      fill = fct_relevel(Disease_Group,
                         c("Control", "PD", "PDD"))
    )
  ) + 
  geom_boxplot() +
  facet_grid(~ dataset) +
  scale_fill_manual(values = c("#868686", "#c4ebe3", "#73cab9", "#19967d")) +
  labs(x = "Cell type", y = "Cell type proportion", fill = "Disease group") + 
  theme_rhr


```
- Qualitatively, it is hard to draw any hard conclusions from the above plot, but might tentatively say:
    - Overall, proportions of each cell type somewhat comparable between the two datasets, albeit with some differences e.g. excitatory neuron proportions look higher in SRP058181 compared to own data, while vascular proportions in SRP058181 look lower than in our data. 
    - Astrocyte and oligodendrocyte proportions appear to be quite similar between the two datasets.
    - Similar rise of microglial and vascular proportions in PD compared to controls seen in both datasets.
    
# Conclusion
- Looking only at controls, the two datasets are indistinguishable, suggesting that cell-type proportions between the two are similar.
- We do pick up on some similar themes e.g. astrocyte and oligodendrocyte proportions appear to be quite similar between the two datasets; similar rise of microglial and vascular proportions in PD compared to controls seen in both datasets.
- Thus, would consider this somewhat of a validation of our deconvolution results.

# Session info
```{r}
sessionInfo()
```

